iothub/device/src/Pipeline/AmqpTransportHandler.cs

```diff
@@ -492,43 +496,60 @@ private void TwinMessageListener(AmqpMessage responseFromService, string correla
-                            task.TrySetException(ex);
+                            pendingOperation.CompletionTask.TrySetException(ex);
                         }
                     }
                     else
                     {
                         // This can happen if we received a message from service with correlation Id that was not set by SDK or does not exist in dictionary.
                         if (Logging.IsEnabled)
-                            Logging.Info("Could not remove correlation Id to complete the task awaiter for a twin operation.", nameof(TwinMessageListener));
+                            Logging.Info(this, "Could not remove correlation Id to complete the task awaiter for a twin operation.", nameof(TwinMessageListener));
                     }
                 }
                 else if (correlationId.StartsWith(AmqpTwinMessageType.Put.ToString(), StringComparison.OrdinalIgnoreCase))
                 {
                     // This is an acknowledgement received from service for subscribing to desired property updates
                     if (Logging.IsEnabled)
-                        Logging.Info("Subscribed for twin desired property updates successfully", nameof(TwinMessageListener));
+                        Logging.Info(this, "Subscribed for twin desired property updates successfully", nameof(TwinMessageListener));
                 }
             }
         }
 
-        private void RemoveOldOperations(object _)
+        private void RemoveOldOperations(object state)
         {
-            _ = _twinResponseTimeouts
-                .Where(x => DateTimeOffset.UtcNow - x.Value > s_twinResponseTimeout)
+            if (state is not TimeSpan maxAge)
+            {
+                maxAge = s_twinResponseTimeout;
+            }
+
+            if (Logging.IsEnabled)
+                Logging.Info(this, $"Removing operations older than {maxAge}", nameof(RemoveOldOperations));
+
+            int canceledOperations = _pendingTwinOperations
+                .Where(x => DateTimeOffset.UtcNow - x.Value.RequestSentOnUtc > maxAge)
                 .Select(x =>
                     {
-                        _twinResponseCompletions.TryRemove(x.Key, out TaskCompletionSource<AmqpMessage> _);
-                        _twinResponseTimeouts.TryRemove(x.Key, out DateTimeOffset _);
+                        if (_pendingTwinOperations.TryRemove(x.Key, out PendingAmqpTwinOperation pendingOperation))
+                        {
+                            if (Logging.IsEnabled)
+                                Logging.Error(this, $"Removing twin response for {x.Key}", nameof(RemoveOldOperations));
+
+                            pendingOperation.CompletionTask.TrySetException(new IotHubClientException("Did not receive twin response from service.", IotHubClientErrorCode.NetworkErrors));
+                        }
                         return true;
-                    });
+                    })
+                .Count();
```

Comment:
The code in select wasn't executing because the IQueryable wasn't finalized. So we'll get the count and print it, and it forces execution of the lambda.